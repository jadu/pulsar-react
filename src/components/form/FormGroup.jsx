import React from 'react';
import classnames from 'classnames';
import shortid from 'shortid';

import Fieldset from './Fieldset';
import FormLabel from './FormLabel';
import ErrorBlock from './ErrorBlock';
import HelpBlock from './HelpBlock';

export default class FormGroup extends React.Component {

  static defaultProps = {
    className: 'form__group',
    required: false
  };

  componentWillMount() {
    this.guid = shortid.generate();
  }

  render() {
    if (!this.props.children) {
      return null;
    }

    let {
      appendText,
      buttonGroup,
      changed,
      checkbox,
      children,
      choiceGroup,
      className,
      error,
      flushLabel,
      helpText,
      indented,
      inlineCheckbox,
      inlineRadioButton,
      labelText,
      prependText,
      required,
      radio,
      success,
      toggle,
      topLabel,
      warning,
      width,
      ...props} = this.props;

    // Convert variants to their required classes
    let variantClasses = classnames(className, {
      [`form__control-col--${width}`]: width,
      'form__button-group': buttonGroup,
      'form-choice': choiceGroup,
      'form__group--indent': indented,
      'form__group--toggle': toggle,
      'form__group--top': topLabel,
      'form__group--flush': flushLabel,
      'form-checkbox': checkbox,
      'form-checkbox-inline': inlineCheckbox,
      'form-radio-inline': inlineRadioButton,
      'form-radio': radio,
      'has-changed': changed,
      'has-success': success,
      'has-warning': warning,
      'has-error': error
    });

    // GUID to use if no explicit ID has been set
    let idGuid = 'id-guid-' + this.guid;
    
    // GUID to use to link the input with any help text 
    let helpGuid = 'help-guid-' + this.guid;
    
    // GUID to use to link the input with any errors 
    let errorGuid = 'error-guid-' + this.guid;

    // Build list of IDs for the input based on the presence of help/error text 
    let ariaDescribedby = classnames({
      [`${errorGuid}`]: error,
      [`${helpGuid}`]: this.props.helpText
    });

    // If help/errors are present pass the GUIDs to the child inputs
    let childrenWithGuids = React.cloneElement(children, {
      ariaDescribedby: ariaDescribedby ? ariaDescribedby : null,
      idGuid: idGuid
    });

    // Build list of classNames to be used on the controls block
    let controlsClassName = classnames('controls', {
      'btn__group': buttonGroup
    });

    // Form controls
    let inputBlock = (
      <>{childrenWithGuids}</>
    );

    // Wrap with input group if appended/prepended text is present
    if (appendText || prependText) {
      inputBlock = (
        <div className="input-group">
          {prependText && 
            <span className="input-group-addon">{prependText}</span>
          }
          {childrenWithGuids}
          {appendText && 
            <span className="input-group-addon">{appendText}</span>
          }
        </div>
      );
    }

    // Wrap input block alongside errors and help
    let controlsBlock = (
      <div className={controlsClassName}>
        {inputBlock}
        <ErrorBlock errorGuid={errorGuid}>
          {error}
        </ErrorBlock> 
        <HelpBlock helpGuid={helpGuid}>
          {helpText}
        </HelpBlock>
      </div>
    );

    // Standard group markup strategy for most components
    let groupBlock = (
      <div className={variantClasses}>
        <FormLabel 
          required={required} 
          idGuid={idGuid}
        >
          {labelText}
        </FormLabel>
        {controlsBlock}
      </div>
    );

    // Switch markup strategies for certain non-standard components
    if (choiceGroup) {
      groupBlock = (
        <Fieldset 
          className={variantClasses}
          legendText={labelText}
          legendClassName="control__label"
        >
          {controlsBlock}
        </Fieldset>
      );
    }
    else if (toggle) {
      groupBlock = (
        <div className={variantClasses}>
          <FormLabel 
            className="toggle-switch-wrapper-label" 
            required={required} 
            idGuid={idGuid}
          >
            <span className="control__label">
              {labelText}
            </span>
            {controlsBlock}
          </FormLabel>
        </div>
      );
    } 
    else if (inlineCheckbox || inlineRadioButton) {
      groupBlock = (
        <div className={variantClasses}>
          {controlsBlock}
          {labelText}
        </div>
      );
    }

    return (
      <>{groupBlock}</>
    );
  }
}