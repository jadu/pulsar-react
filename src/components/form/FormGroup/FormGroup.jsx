import React from 'react';
import classnames from 'classnames';
import shortid from 'shortid';

import Fieldset from '../Fieldset/Fieldset';
import FormLabel from '../FormLabel/FormLabel';
import ErrorBlock from '../ErrorBlock/ErrorBlock';
import HelpBlock from '../HelpBlock/HelpBlock';

export default class FormGroup extends React.Component {

  static defaultProps = {
    className: 'form__group',
    required: false
  };

  render() {
    if (!this.props.children) {
      return null;
    }

    let {
      appendText,
      buttonGroup,
      changed,
      checkbox,
      children,
      choiceGroup,
      className,
      controls,
      error,
      flushLabel,
      helpText,
      hideLabel,
      indented,
      inline,
      inlineCheckbox,
      inlineRadioButton,
      labelText,
      prependText,
      required,
      radioButton,
      rangeInput,
      success,
      toggle,
      topLabel,
      warning,
      width,
      ...props} = this.props;

    // Convert variants to their required classes
    let variantClasses = classnames(className, {
      [`form__control-col--${width}`]: width,
      'form__button-group': buttonGroup,
      'form-choice': choiceGroup,
      'form__group--compound': inline,
      'form__group--indent': indented,
      'form__group--toggle': toggle,
      'form__group--top': topLabel,
      'form__group--flush': flushLabel,
      'form-checkbox': checkbox,
      'form-checkbox-inline': inlineCheckbox,
      'form-radio-inline': inlineRadioButton,
      'form-radio': radioButton,
      'form-range': rangeInput,
      'has-changed': changed,
      'has-success': success,
      'has-warning': warning,
      'has-error': error
    });

    let guid = this.props.guid || shortid.generate();

    // GUID to use if no explicit ID has been set
    let idGuid = 'id-guid-' + guid;
    
    // GUID to use to link the input with any help text 
    let helpGuid = 'help-guid-' + guid;
    
    // GUID to use to link the input with any errors 
    let errorGuid = 'error-guid-' + guid;

    // Build list of IDs for the input based on the presence of help/error text 
    let ariaDescribedby = classnames({
      [`${errorGuid}`]: error,
      [`${helpGuid}`]: this.props.helpText
    });

    // If help/errors are present pass the GUIDs to the child inputs
    let childrenWithGuids = React.Children.map(
      children,
      (child, i) => {
        if (['div', 'p'].indexOf(child.type) !== -1) {
          return child;
        }

        return React.cloneElement(child, {
          ariaDescribedby: ariaDescribedby ? ariaDescribedby : null,
          id: idGuid
        });
      }
    );

    // Build list of classNames to be used on the controls block
    let controlsClassName = classnames('controls', {
      'btn__group': buttonGroup
    });

    // Form controls
    let inputBlock = (
      <React.Fragment>{childrenWithGuids}</React.Fragment>
    );

    // Wrap with input group if appended/prepended text is present
    if (appendText || prependText) {
      inputBlock = (
        <div className="input-group">
          {prependText && 
            <span className="input-group-addon">{prependText}</span>
          }
          {childrenWithGuids}
          {appendText && 
            <span className="input-group-addon">{appendText}</span>
          }
        </div>
      );
    }

    // Wrap input block alongside errors and help
    let controlsBlock = (
      <div className={controlsClassName}>
        {inputBlock}
        <ErrorBlock errorGuid={errorGuid}>
          {error}
        </ErrorBlock> 
        <HelpBlock helpGuid={helpGuid}>
          {helpText}
        </HelpBlock>
      </div>
    );

    // Standard group markup strategy for most components
    let groupBlock = (
      <div className={variantClasses}>
        {labelText &&
        <FormLabel 
          required={required} 
          idGuid={idGuid}
          tag={controls ? 'span' : 'label'}
          hideLabel={hideLabel}
        >
          {labelText}
        </FormLabel>
        }

        {controlsBlock}
      </div>
    );

    // Switch markup strategies for certain non-standard components
    if (choiceGroup) {
      groupBlock = (
        <Fieldset 
          className={variantClasses}
          legendText={labelText}
          legendClassName="control__label"
        >
          {controlsBlock}
        </Fieldset>
      );
    }
    else if (toggle) {
      groupBlock = (
        <div className={variantClasses}>
          <FormLabel 
            className="toggle-switch-wrapper-label" 
            required={required} 
            idGuid={idGuid}
          >
            <span className="control__label">
              {labelText}
            </span>
            {controlsBlock}
          </FormLabel>
        </div>
      );
    } 
    else if (inlineCheckbox || inlineRadioButton) {
      groupBlock = (
        <div className={variantClasses}>
          {controlsBlock}
          {labelText}
        </div>
      );
    }

    return (
      <React.Fragment>{groupBlock}</React.Fragment>
    );
  }
}