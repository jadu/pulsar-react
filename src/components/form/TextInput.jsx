import React from 'react';

export default class TextInput extends React.Component {
  render() {
    let {
      ariaDescribedby,
      id,
      idGuid,
      ...props
    } = this.props;

    return (
      <input 
        id={id ? id : idGuid} 
        className="form__control" 
        type="text" 
        aria-describedby={ariaDescribedby}
        {...props} 
      />
    );
  }
}