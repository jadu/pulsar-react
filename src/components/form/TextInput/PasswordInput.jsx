import React from 'react';

export default class PasswordInput extends React.Component {
  render() {
    let {
      helpGuid,
      helpText,
      id,
      idGuid,
      ...props
    } = this.props;

    return (
      <input 
        id={id ? id : idGuid} 
        className="form__control" 
        type="password" 
        aria-describedby={helpText && helpGuid} 
        {...props} 
      />
    );
  }
}