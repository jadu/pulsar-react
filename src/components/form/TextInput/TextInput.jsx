import React from 'react';
import FormLabel from '../FormLabel/FormLabel';

export default class TextInput extends React.Component {
  render() {
    let {
      helpText,
      id,
      labelText,
      ...props
    } = this.props;

    let helpClasses = 'help-block';

    let helpBlock = helpText ? (
      <span className={helpClasses}>{helpText}</span>
    ) : null;

    return (
      <>
      <FormLabel id={id} labelText={labelText} required={props.required} />
      <div className="controls">
        <input className="form__control" type="text" {...props} />
        {helpBlock}
      </div>
      </>
    );
  }
}