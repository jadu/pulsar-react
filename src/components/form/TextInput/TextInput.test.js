'use strict';

import React from 'react';
import { shallow, mount } from 'enzyme';


import FormGroup from '../FormGroup';
import TextInput from './TextInput';

it('renders a form group with text input', () => {
  const tree = shallow(
    <FormGroup labelText="Basic">
      <TextInput placeholder="foo" />
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});

it('renders the basic input', () => {
  const tree = shallow(
    <TextInput />
  );
  expect(tree).toMatchSnapshot();
});

it('renders the basic input with id', () => {
  const tree = shallow(
    <TextInput id="foo" />
  );
  expect(tree).toMatchSnapshot();
});

it('renders the basic input with required', () => {
  const tree = shallow(
    <TextInput required />
  );
  expect(tree).toMatchSnapshot();
});

it('renders the basic input with placeholder', () => {
  const tree = shallow(
    <TextInput placeholder="foo" />
  );
  expect(tree).toMatchSnapshot();
});




