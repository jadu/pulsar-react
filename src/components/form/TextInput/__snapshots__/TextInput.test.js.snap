// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a form group with text input 1`] = `
<Fragment>
  <div
    className="form__group"
  >
    <FormLabel
      className="control__label"
      idGuid="id-guid-1"
      required={false}
    >
      Basic
    </FormLabel>
    <div
      className="controls"
    >
      <TextInput
        ariaDescribedby={null}
        className="form__control"
        idGuid="id-guid-1"
        placeholder="foo"
        type="text"
      />
      <ErrorBlock
        errorGuid="error-guid-1"
      />
      <HelpBlock
        helpGuid="help-guid-1"
      />
    </div>
  </div>
</Fragment>
`;

exports[`renders the basic input 1`] = `
<input
  className="form__control"
  type="text"
/>
`;

exports[`renders the basic input with id 1`] = `
<input
  className="form__control"
  id="foo"
  type="text"
/>
`;

exports[`renders the basic input with placeholder 1`] = `
<input
  className="form__control"
  placeholder="foo"
  type="text"
/>
`;

exports[`renders the basic input with required 1`] = `
<input
  className="form__control"
  required={true}
  type="text"
/>
`;
